File: Solutions/Minimum_Number_of_Arrows_to_Burst_Balloons/Minimum_Number_of_Arrows_to_Burst_Balloons.js
var findMinArrowShots = function(points) {
    if (points.length === 1)
        return 1;
    
    let res = points.length;
    
    points.sort((a,b) => a[0]-b[0] || (a[1]-a[0])-(b[1]-b[0]));
    
    for (let i = 1; i < points.length; i++) {
        if (points[i-1][0] <= points[i][0] && points[i][0] <= points[i-1][1]) {
            res--;
            [points[i][0], points[i][1]] = [Math.max(points[i-1][0], points[i][0]), Math.min(points[i-1][1], points[i][1])];
        }
    }
    
    return res;
};

//Third fastest
/**
 * @param {number[][]} points
 * @return {number}
 */
var findMinArrowShots = function(points) {
    if (points.length == 0) {
      return 0;
    }
      
    points.sort((a, b) => a[1] - b[1]);
   let arrowPos = points[0][1];
   let arrowCnt = 1;
   for (let i = 1; i < points.length; i++) {
       if (arrowPos >= points[i][0]) {
           continue;
       }
       arrowCnt++;
       arrowPos = points[i][1];
   }
   return arrowCnt;
};

//Second fastest
/**
 * @param {number[][]} points
 * @return {number}
 */
var findMinArrowShots = function(points) {
    if (!points.length) return 0
    points.sort((a, b) => a[1] - b[1])
    let arrow = 1, lastEnd = points[0][1]
    for (let i = 1; i < points.length; i++) {
      let [newStart, newEnd] = points[i]
      if (lastEnd < newStart) {
        arrow++
        lastEnd = newEnd
      }
    }
    return arrow
  };

//Fastest
/**
 * @param {number[][]} points
 * @return {number}
 */
var findMinArrowShots = function(points) {
    if (points.length === 0) {
        return 0;
    }
    points.sort((a, b) => {
        return a[1] - b[1];
    })
    
    let min = points[0][0];
    let max = points[0][1];
    let arrow = 0;
    const length = points.length;
    
    for (let i = 1; i < length; i++) {
        if (points[i][0] <= max) {
            min = min > points[i][0]? min:points[i][0];
            max = max > points[i][1]? points[i][1]:max;
            continue;
        }
        arrow += 1;
        min = points[i][0];
        max = points[i][1];
    }
    
    arrow++;
    return arrow;
};

File: Solutions/Minimum_Number_of_Arrows_to_Burst_Balloons/Minimum_Number_of_Arrows_to_Burst_Balloons.cpp
    bool comp(const vector<int> &t1,const vector<int> &t2)
{
	return (t1[1]<t2[1]) ;
}
class Solution {

public:
    int findMinArrowShots(vector<vector<int>>& points) {
    
    
    if(points.size()==0) return 0;
    
    if(points.size()==1) return 1;
    
    
    sort(points.begin(),points.end(),comp);
    
    int count=1;
    int prev=points[0][1];
    
    for(int i=1;i<points.size();i++)
    {
        if(points[i][0]<=prev) continue; // overlapping part
        
        count++;
        prev=points[i][1];
    }
    
    //++arrow_cnt;//for last arrow
    return count;
    
}
};

//Second fastest
class Solution {
public:
    static bool comp (pair<int, int>& i, pair<int, int>& j) { 
        return i.second < j.second; 
    }
    
    int findMinArrowShots(vector<vector<int>>& p) {
        ios::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);
        size_t n = p.size();
        if (n == 0)
            return 0;
        vector<pair<int, int>> interval;
        for(int i = 0; i < n; ++i){
            interval.push_back({p[i][0], p[i][1]});
        }
        sort(interval.begin(), interval.end(), comp);
        
        int res = 1;
        int ce = interval[0].second;
        for (int i = 1; i < n; ++i) {
            if (interval[i].first <= ce)
            {
                continue;
            } else {
                ce = interval[i].second;
                ++res;    
            }
                
            
        }
        return res;
    }
};

//Fastest
class Solution {
public:
    static bool comp (pair<int, int>& i, pair<int, int>& j) { 
        return i.second < j.second; 
    }
    
    int findMinArrowShots(vector<vector<int>>& p) {
        ios::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);
        auto n = p.size();
        if (n == 0)
            return 0;
        int i;
        vector<pair<int, int>> interval;
        for(i = 0; i < n; ++i){
            interval.push_back({p[i][0], p[i][1]});
        }
        sort(interval.begin(), interval.end(), comp);
        
        auto res = 1;
        auto ce = interval[0].second;
        for (i = 1; i < n; ++i) {
            if (interval[i].first <= ce)
            {
                continue;
            } else {
                ce = interval[i].second;
                ++res;    
            }
                
            
        }
        return res;
    }
};